src/darker/tests/test_highlighting.py

```diff
@@ -1,50 +1,174 @@
 """Unit tests for :mod:`darker.highlighting`"""
 
+import os
 import sys
+from pathlib import Path
 from unittest.mock import Mock, patch
 
 import pytest
 from pygments.token import Token
 
-from darker.highlighting import lexers, with_pygments, without_pygments
+from darker.command_line import parse_command_line
+from darker.highlighting import colorize, lexers, should_use_color
 
 
-def test_colorize_import_without_pygments():
-    """Dummy ``colorize()`` is used if Pygments isn't available"""
-    modules = sys.modules.copy()
-    del modules["darker.highlighting"]
-    # cause an ImportError for `import pygments`:
-    modules["pygments"] = None  # type: ignore[assignment]
-    with patch.dict(sys.modules, modules, clear=True):
-        # pylint: disable=import-outside-toplevel
-
-        from darker.highlighting import colorize
-
-        assert colorize == without_pygments.colorize
-
-
-def test_colorize_import_with_pygments():
-    """The real ``colorize()`` is used if Pygments is available"""
-    assert "pygments" in sys.modules
+@pytest.mark.parametrize(
```

Comment:
[Suggestion] You can use 2 or more parametrize decorator to create a matrix here instead of manually inputting the params.

Reply:
@roniemartinez thanks for the suggestion!

I actually prefer the explicit, human-readable list of parameter combinations and expected behavior. This way it's easy to read and understand what the behavior is and should be.

Also, the expected value (`USE_COLOR` or not) depends on the input values in a non-trivial way. If the test function derived it from values of parameters in the matrix, it would need to re-implement the function being tested.

In general, I've grown to appreciate the principle that DRY ("don't repeat yourself") doesn't apply to test code – instead test code should strive for maximum simplicity, clarity and understandability. Otherwise there's a risk that there's a bug not only in the code being tested but also the complicated test code.

In this particular case, it would be nice if the initialization of environment variables, `pyproject.toml`, command line options and the availability of the `pygments` module was simpler while still preserving the readability of input values and the expected value, but I couldn't find a better solution than this.

Reply:
Actually, since `should_use_color()` returns `False` always if `pygments` can't be imported, that half of the test cases _could_ actually be separated into its own test case with matrix parameterization. I'll experiment with that.

Reply:
I favor the suggestion of @roniemartinez. A human-readable list would be preffered, when it is a short list. This is a very long list and therefore not beneficial IMO.

Reply:
@roniemartinez, I now converted the non-`pygments` half of these test cases into a parameterized matrix test. I also touched the remaining half a bit to hopefully make it even more understandable but also more concise.

What do you think?

Reply:
@MatthijsBurgh, how would you implement it so that the rules for determining the expected `True`/`False` value don't just add a second (and similarly bug-risky) implementation of the algorithm which is being tested?

Also, what do you think about the length of the list of test cases after I moved half of them into a parameterized matrix test?

Reply:
I guess two thirds of the remaining test cases could be extracted into two separate test functions for the cases with the `--color` and `--no-color` command-line options, which always force highlighting on or off, respectively, if `pygments` is installed.

But now, instead of one long easily eye-scannable list of test cases, we'd have four different test functions, increasing the chance of mistakes in the test code, and making it more challenging to read and understand the tests when debugging possible future failures.

So I'm not 100% sure this is a good idea either. Please convince me otherwise!

Reply:
Does pytest show the chosen values for failing runs? If that is the case, I think we should go for the matrix.

> But now, instead of one long easily eye-scannable list of test cases, we'd have four different test functions

Do you mean four different functions or decorators?

Reply:
@akaihola Just to add clarity into the concept of the parametrize matrix. I think this is more readable.

```python
@pytest.mark.parametrize("tty", ["tty", ""])
@pytest.mark.parametrize("arg", ["--color", ""])
@pytest.mark.parametrize("env", ["1", "0"])
@pytest.mark.parametrize("color", [True, False])
def test_parametrized_matrix(color, env, arg, tty):
    assert True
``` 

This in turn generates 16 tests

```
collected 16 items                                                                                                               

tests/test_sample.py::test_parametrized_matrix[True-1---color-tty] PASSED                                                  [  6%]
tests/test_sample.py::test_parametrized_matrix[True-1---color-] PASSED                                                     [ 12%]
tests/test_sample.py::test_parametrized_matrix[True-1--tty] PASSED                                                         [ 18%]
tests/test_sample.py::test_parametrized_matrix[True-1--] PASSED                                                            [ 25%]
tests/test_sample.py::test_parametrized_matrix[True-0---color-tty] PASSED                                                  [ 31%]
tests/test_sample.py::test_parametrized_matrix[True-0---color-] PASSED                                                     [ 37%]
tests/test_sample.py::test_parametrized_matrix[True-0--tty] PASSED                                                         [ 43%]
tests/test_sample.py::test_parametrized_matrix[True-0--] PASSED                                                            [ 50%]
tests/test_sample.py::test_parametrized_matrix[False-1---color-tty] PASSED                                                 [ 56%]
tests/test_sample.py::test_parametrized_matrix[False-1---color-] PASSED                                                    [ 62%]
tests/test_sample.py::test_parametrized_matrix[False-1--tty] PASSED                                                        [ 68%]
tests/test_sample.py::test_parametrized_matrix[False-1--] PASSED                                                           [ 75%]
tests/test_sample.py::test_parametrized_matrix[False-0---color-tty] PASSED                                                 [ 81%]
tests/test_sample.py::test_parametrized_matrix[False-0---color-] PASSED                                                    [ 87%]
tests/test_sample.py::test_parametrized_matrix[False-0--tty] PASSED                                                        [ 93%]
tests/test_sample.py::test_parametrized_matrix[False-0--] PASSED                                                           [100%]
```

Reply:
Pytest does show the chosen values for failing runs.

I meant four different test functions if I split out the "simple" cases:
- `pygments` not installed -> always `False`
- `pygments` installed and `--color` used -> always `True`
- `pygments` installed and `--no-color` used -> always `False`
- `pygments` installed and no command line option -> depends on `pyproject.toml` and environment variables; This would be either a parametrized list of test cases or a matrix.

But if we go with a matrix either for that 4th set of remaining test cases, or for the whole shebang, my question remains:
>How would you implement the test so that the rules for determining the expected `True`/`False` value don't just add a second (and similarly bug-risky) implementation of the algorithm which is being tested (i.e. the logic in `should_use_color()`)?

Reply:
@roniemartinez thanks for the clear code example.

However, you haven't implemented the assertion yet in the example, and that's precisely the tricky part. The expected return value from `should_use_color()` _depends on the input values in a complex way_.

How do you handle that in a matrix parameterization without re-implementing `should_use_color()`'s logic in the test function?

(this is essentially the same question as in me previous comment, reworded)

Reply:
I see `--no-color` forces everything to not use color. I would create a separate test for that to simplify it. As for the other conditions, they seem to be just an "OR" condition. I would say 2 tests with matrices. Please correct the inaccuracy.

```python
# None means not present


@pytest.mark.parametrize("tty", ["tty", None])
@pytest.mark.parametrize("env", [1, 0, None])
@pytest.mark.parametrize("color", [True, False, None])
def test_not_use_color(color, env, tty):
    # with --no-color should always disable color
    ...
    assert should_use_color() is False

@pytest.mark.parametrize("tty", ["tty", None])
@pytest.mark.parametrize("arg", ["--color", None])
@pytest.mark.parametrize("env", [1, 0, None])
@pytest.mark.parametrize("color", [True, False, None])
def test_matrix(color, env, arg, tty):
    # at this point `color or env or arg or tty` should determine the output
    ...
    assert should_use_color() == color or env or arg or tty
```

Reply:
>`--no-color` forces everything to not use color. I would create a separate test for that to simplify it

Yes, that's probably a decent idea. And it's what I proposed earlier above, and I'm currently experimenting with it in #356 (see commit db57c55). It does lead to repetition of test setup code, and I'll probably refactor it into fixtures.

>other conditions, they seem to be just an "OR" condition

Yes, and I argue that writing `color or env or arg or tty` is just a contracted re-implementation of the combination of `parse_command_line()` and `should_use_color()`. The latter, for example, has:
```python
    if config_color is not None:
        use_color = config_color
    else:
        use_color = sys.stdout.isatty()

    if use_color:
        try:
            import pygments  # noqa

            return True
        except ImportError:
            pass
    return False
```
So if we make an aberration (or a typo, or whatever) in the "OR" clause, that same aberration might exist in this code which we're testing.

I still stand behind my principle of avoiding re-implementation of the logic being tested.

In other words, my rule of thumb would be:

>_A test case should declare the expected value, not calculate it._

Reply:
Anyway, I'm going to refine the tests in #356 since the addition of `NO_COLOR` again doubles the number of test cases and calls for refactoring of the tests.

I'll include both #353 and #356 in [1.5.0](https://github.com/akaihola/darker/milestone/13), so we can continue work on the test suite there.

Big thanks to you both, @roniemartinez and @MatthijsBurgh, for working on this with me!